File: server/app/matcha/models/insurance.py
"""Carrier quote/bind request + response models (Coterie)."""

from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

# Lines Coterie small-commercial can quote, expressed as limit_adequacy line keys
# so a bound policy upserts cleanly into company_coverage_lines. bop = a Business
# Owner's Policy (packaged GL + property); kept distinct so the UI can offer it.
QuotableLine = Literal["bop", "gl", "wc", "professional"]


class QuoteRequest(BaseModel):
    """Caller confirms/edits the prefilled inputs, then submits for a quote."""

    line: QuotableLine
    # Everything below defaults from the company's own data via
    # coterie_service.build_quote_request; the caller may override before submit.
    legal_name: Optional[str] = Field(None, max_length=255)
    naics: Optional[str] = Field(None, max_length=10)
    state: Optional[str] = Field(None, max_length=2)
    zip_code: Optional[str] = Field(None, max_length=10)
    headcount: Optional[int] = Field(None, ge=0)
    annual_payroll: Optional[float] = Field(None, ge=0)
    annual_revenue: Optional[float] = Field(None, ge=0)


class QuoteResponse(BaseModel):
    id: str
    line: str
    carrier: str
    status: str
    quote_ref: Optional[str] = None
    premium_cents: Optional[int] = None
    expires_at: Optional[str] = None
    error_message: Optional[str] = None
    certificate_id: Optional[str] = None
    created_at: Optional[str] = None


# --- broker-placed quoting (Broker Quoting Desk) -------------------------------

class BrokerQuoteRequest(QuoteRequest):
    """A broker requesting a quote on behalf of a client in their book. Same
    prefill/override fields as a client self-serve quote, plus placement metadata
    the broker sets (commission + a private note)."""

    commission_bps: Optional[int] = Field(None, ge=0, le=10_000)
    broker_note: Optional[str] = Field(None, max_length=2_000)


class PresentRequest(BaseModel):
    """Broker presents a quoted policy to the on-platform client for acceptance."""

    commission_bps: Optional[int] = Field(None, ge=0, le=10_000)
    broker_note: Optional[str] = Field(None, max_length=2_000)


class FnolRequest(BaseModel):
    """File a First Notice of Loss to the carrier from a logged IR incident."""

    incident_id: UUID
    description: Optional[str] = Field(None, max_length=4_000)

File: client/src/api/risk/insurance.ts
import { api } from '../client'

export type QuotableLine = 'bop' | 'gl' | 'wc' | 'professional'
// 'presented' = a broker placed this quote and presented it for the client to accept.
export type QuoteStatus = 'draft' | 'quoted' | 'presented' | 'bound' | 'expired' | 'error'

export type QuotePayload = {
  product: string
  line: string
  business: { legal_name: string | null; naics: string | null; industry: string | null; state: string | null; zip: string | null }
  exposure: { headcount: number | null; annual_payroll: number | null; annual_revenue: number | null }
}

export type Prefill = { line: string; payload: QuotePayload; mock_mode: boolean }

export type Quote = {
  id: string
  line: string
  carrier: string
  status: QuoteStatus
  quote_ref: string | null
  premium_cents: number | null
  expires_at: string | null
  error_message: string | null
  certificate_id: string | null
  created_at: string | null
  // present when a broker placed the quote (broker-in-the-middle path)
  placement?: string
  presented_at?: string | null
  commission_bps?: number | null
  broker_note?: string | null
}

export type QuoteInput = {
  line: QuotableLine
  legal_name?: string | null
  naics?: string | null
  state?: string | null
  zip_code?: string | null
  headcount?: number | null
  annual_payroll?: number | null
  annual_revenue?: number | null
}

export function getPrefill(line: QuotableLine) {
  return api.get<Prefill>(`/insurance/prefill?line=${line}`)
}
export function listQuotes() {
  return api.get<{ quotes: Quote[] }>('/insurance/quotes')
}
export function createQuote(input: QuoteInput) {
  return api.post<Quote>('/insurance/quote', input)
}
export function bindQuote(id: string) {
  return api.post<Quote>(`/insurance/quotes/${id}/bind`)
}
